File: python/7 - Trees/450.DeleteNodeinaBST.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# class Solution(object):
#     def deleteNode(self, root, key):
#         """
#         :type root: Optional[TreeNode]
#         :type key: int
#         :rtype: Optional[TreeNode]
#         """
#         if not root:
#             return root
#         if root.val > key:
#             root.left = self.deleteNode(root.left, key)
#         elif root.val < key:
#             root.right = self.deleteNode(root.right, key)
#         else:
#             if not root.left:
#                 return root.right
#             elif not root.right:
#                 return root.left
#             else:
#                 cur = root.right
#                 while cur.left:
#                     cur = cur.left
#                 root.val = cur.val
#                 root.right = self.deleteNode(root.right, cur.val)
#         return root


class Solution(object):
    def deleteNode(self, root, key):
        """
        :type root: Optional[TreeNode]
        :type key: int
        :rtype: Optional[TreeNode]
        """
        if not root:
            return root
        parent = None
        cur = root
        while cur and cur.val != key:
            parent = cur
            if key > cur.val:
                cur = cur.right
            else:
                cur = cur.left
        if not cur:
            return root
        if not cur.left or not cur.right:
            child = cur.left if cur.left else cur.right
            if not parent:
                return child
            if parent.left == cur:
                parent.left = child
            else:
                parent.right = child
        else:
            par = None
            delNode = cur
            cur = cur.right
            while cur.left:
                par = cur
                cur = cur.left
            if par:
                par.left = cur.right
                cur.right = delNode.right
            cur.left = delNode.left
            if not parent:
                return cur
            if parent.left == delNode:
                parent.left = cur
            else:
                parent.right = cur
        return root

File: ts/7 - Trees/450.DeleteNodeinaBST.ts
/* 
Given a root node reference of a BST and a key, delete the node with the given key in the BST. Return the root node reference (possibly updated) of the BST.

Basically, the deletion can be divided into two stages:

Search for a node to remove.
If the node is found, delete the node.
*/

/* ------------------------------------------------------------------------------- */

class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = val === undefined ? 0 : val;
        this.left = left === undefined ? null : left;
        this.right = right === undefined ? null : right;
    }
}

function deleteNode(root: TreeNode | null, key: number): TreeNode | null {
    if (!root) return root;

    if (root.val > key) {
        root.left = deleteNode(root.left, key);
    } else if (root.val < key) {
        root.right = deleteNode(root.right, key);
    } else {
        // found key and delete it
        if (!root.left) return root.right;
        else if (!root.right) return root.left;
        else {
            // if both left and right subtree exist, replace the target node with the minimum value of right subtree
            let current = root.right;
            while (current.left) {
                current = current.left;
            }
            root.val = current.val;

            // then delete that value from right subtree by recursively calling deleteNode
            root.right = deleteNode(root.right, root.val);
        }
    }

    return root;
}

/*
T.C.: O(log(N))
S.C.: O(H)
*/

/* ------------------------------------------------------------------------------- */
